crates/integration_test/src/tests/readonly_paths/readonly_paths.rs

```diff
@@ -0,0 +1,275 @@
+
+    let spec = get_spec(ro_paths);
+
+    test_inside_container(spec, &|bundle| {
+        use std::{fs, io};
+        let bundle_path = bundle.as_ref();
+        let test_file = bundle_path.join(ro_symlink);
+
+        match std::os::unix::fs::symlink("../readonly_symlink", &test_file) {
+            io::Result::Ok(_) => { /* This is expected */ }
+            io::Result::Err(e) => {
+                bail!("error in creating symlink, to {:?} {:?}", test_file, e);
+            }
+        }
+        let r_path = match fs::read_link(&test_file) {
+            io::Result::Ok(p) => p,
+            io::Result::Err(e) => {
+                bail!("error in reading symlink at {:?} : {:?}", test_file, e);
+            }
+        };
+
+        match fs::metadata(&r_path) {
+            io::Result::Ok(md) => {
+                bail!(
+                    "reading symlink for {:?} should have given error, found {:?} instead",
+                    test_file,
+                    md
+                )
+            }
+            io::Result::Err(e) => {
+                let err = e.kind();
+                if let io::ErrorKind::NotFound = err {
+                    return Ok(());
+                } else {
+                    bail!("expected not found error, got {:?}", err);
+                }
+            }
+        }
+    })
+}
+
+fn test_node(mode: u32) -> TestResult {
+    let root = PathBuf::from("/");
+    let ro_device = "readonly_device";
+    let ro_paths = vec![root.join(&ro_device).to_string_lossy().to_string()];
+
+    let spec = get_spec(ro_paths);
+
+    test_inside_container(spec, &|bundle| {
+        use std::{fs, io};
+
+        let bundle_path = bundle.as_ref();
+        let test_file = bundle_path.join(&ro_device);
+        // NOTE
+        // yes, I know using unsafe willy-nilly is a bad idea,
+        // especially given that OpenOptionsExt in std::os::unix::fs does provide a method
+        // to set mode in open options, like this :
+
+        // use std::os::unix::fs::OpenOptionsExt;
+        // let mut opts = fs::OpenOptions::new();
+        // opts.mode(mode);
+        // opts.create(true);
+        // if let io::Result::Err(e) = opts.open(&test_file) {
+        //     bail!(
+        //         "could not create device node at {:?} with mode {}, got error {:?}",
+        //         test_file,
+        //         mode ^ 0o666,
+        //         e
+        //     );
+        // }
+
+        // but that gives OsErr 22, invalid arguments.
+        // That is why we directly use mknod from lib here
+
+        let _path = test_file.to_string_lossy().as_ptr() as *const i8;
+        let r = unsafe { libc::mknod(_path, mode, 0) };
+        if r != 0 {
+            bail!(
+                "error in creating a device node at {:?} with mode {:?}, got return code {}",
+                test_file,
+                mode,
+                r
+            );
+        }
+
```

Comment:
The use of unsafe irks me here, as there should be a better way to implement this. I tried setting the mode in open options, but it always gives OsErr 22, invalid argument, as specified in the comment here. Is there a better way to do this, or is there something that I am doing wrong because of which the open option method is failing?

Reply:
Not sure if this is the issue, but can you chain the opts calls? See example: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html

`if let io::Result::Err(e) = fs::OpenOptions::new().mode(mode).create(true).open(&test_file) {...}`

Reply:
Hey, I did try chaining the opts, and that compiled, but still gave the same issue at runtime as the separate value setting.

Reply:
How about: `In order for the file to be created, OpenOptions::write or OpenOptions::append access must be used.` from https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.create

Reply:
Also, can we use nix::mknod instead of libc::mknod?

Reply:
Hey, using `write(true)` fixed the issue :+1:  thanks for the suggestion!
I originally thought `create(true)` by itself would set `write(true)` internally, but we need to do it manually.
This solves the unsafe issue, as well as the separate tests issue :+1:
I also have removed the libc direct dependency and used nix all the way :)